### PR TITLE
cmake refactor

### PR DESCRIPTION
**Pull request by seimtpow**
# Description

Updating cmake to use updated neso-particles (now consumed as submodule) + some other minor updates
added AdaptiveCPP and IntelSYCL as options in FindSYCL
fixed some deprecated code that fails to compile with new intel compiler
updated neso-particles submodule to point to compatible version.

### CODE REVIEW

**Comment by will-saunders-ukaea** [comment]
These fixes for the namespace std includes fix the compiling issues we have been seeing with gcc 11.3.0, 11.4.0 and gcc 12.3.0. (Apparently Nektar saw this as well and a stray "using namespace std" was the cause).

**Comment by seimtpow** [comment]
> These fixes for the namespace std includes fix the compiling issues we have been seeing with gcc 11.3.0, 11.4.0 and gcc 12.3.0. (Apparently Nektar saw this as well and a stray "using namespace std" was the cause)

We should probably remove all the `using namespace <whatever>` in all the headers at some point and then ban it.

File: CMakeLists.txt
Comment (cmacmackin) [inline CMakeLists.txt:59]: Is the need to include this the reason you've switched from finding NESO-Particles as a package to loading it as a subdirectory?

Might it be possible for NESO-Particles to install the restrict-keyword.cmake file and include it in the exported NESO-ParticlesConfig.cmake?

File: CMakeLists.txt
Comment (cmacmackin) [inline CMakeLists.txt:42]: We've spoken about this briefly in person, but I'm not clear on the logic of taking this out of the FindSYCL file.

Comment (cmacmackin) [inline CMakeLists.txt:42]: Presumably this should match whatever you and @will-saunders-ukaea settle on in the [NESO-Particles pull request](https://github.com/ExCALIBUR-NEPTUNE/NESO-Particles/pull/57#discussion_r1515961915).

File: CMakeLists.txt
Comment (will-saunders-ukaea) [reply CMakeLists.txt:59]: Will this disable the ability to use an external NESO-Particles? i.e. I develop NESO-Particles in a completely separate directory (found as a cmake package).

Comment (seimtpow) [reply CMakeLists.txt:59]: No it should still work as a package - it looks like in dev it copies the cmake files across and I lost that in the changes. I'll fix this now.
I'm not sure what the cleanest way to get it to work in a submodule the add_compile_definitions call won't effect the parent scope.
TBH I think this is overkill anyhow, a header with some ifdefs to cover all the compilers - in practice there are only 3 (clang/icpx, gcc, msvc) of them and that's if we care about windows -  and they all recognise __restrict AFAIK anyhow.

File: CMakeLists.txt
Comment (seimtpow) [reply CMakeLists.txt:42]: The more I think about it the more I dislike the idea of FindSYCL *and* the find_sycl macro I wrote - I changed it so I could explicitly ask for a particular SYCL version at "cmake time" and I couldn't think of a standard way to do that with a find_package - its not really a component or a version. 

I think we should just force people to chose - SYCL isn't standardised enough for a find mpi style construct to work well IMO. As I mentioned in another comment it's too easy to get something unexpected this way - setting targets for hipsycl but finding intelSYCL. 

Having said that - what I have here is broken in a couple of different ways so I will fix it.

**Comment by jwscook** [comment]
@cmacmackin would you mind reviewing this and merging if acceptable please?

**Comment by cmacmackin** [comment]
Sorry for not doing this earlier. There kept being new commits pushed so I wasn't sure when it was actually ready.

**Review by cmacmackin — COMMENTED** [review]
When I tried building this I got this error again:
```
   The link interface of target "hdf5-shared" contains:

    MPI::MPI_C

  but the target was not found.  Possible reasons include:
```
I addressed this in ExCALIBUR-NEPTUNE/NESO-Particles#47 but it seems to have been lost in @seimtpow refactoring.

**Review by cmacmackin — APPROVED** [review]
I've made some changes necessary to make everything build on my system (plus address a flaw I noticed). It's probably worth @will-saunders-ukaea or @oparry-ukaea checking everything still builds for them. Might also be best to wait until ExCALIBUR-NEPTUNE/NESO-Particles#64 is merged.

**Comment by oparry-ukaea** [comment]
Shall I merge main into this, resolve the conflict with solvers/SimpleSOL/EquationSystems/SOLWithParticlesSystem.cpp and make sure it builds for me?

**Comment by oparry-ukaea** [comment]
I get 
```
CMake Error at CMakeLists.txt:240 (target_link_libraries):
  Target "nesolib" links to:

    NESO-Particles::NESO-Particles
when configuring
  but the target was not found.  Possible reasons include:
...
```
(before merging from main, and having run `git submodule update --init`)

**Comment by cmacmackin** [comment]
Could you send me your `spack-build-01-cmake-out.txt` file please?

**Comment by oparry-ukaea** [comment]
[spack-build-01-cmake-out.txt](https://github.com/ExCALIBUR-NEPTUNE/NESO/files/14957225/spack-build-01-cmake-out.txt)

**Comment by cmacmackin** [comment]
> [spack-build-01-cmake-out.txt](https://github.com/ExCALIBUR-NEPTUNE/NESO/files/14957225/spack-build-01-cmake-out.txt)

That looks like  it ran successfully. I'm not seeing any errors about failing to find NESO-Particles.

**Comment by oparry-ukaea** [comment]
> That looks like it ran successfully. I'm not seeing any errors about failing to find NESO-Particles.

Sorry, was configuring with
`spack build-env neso%gcc cmake -B ...`
and hadn't re-spack-installed NESO-Particles ... :man_facepalming:

Builds with gcc for me, with Will's latest changes plus merge from main
Trying oneapi now, by I assume that'll be fine

**Comment by will-saunders-ukaea** [comment]
Right my understanding is that @seimtpow is doing a cleanup of the NP cmake that should not change functionality. Then we can look to actually merge.

**Comment by oparry-ukaea** [comment]
Builds with oneapi for me too

**Comment by cmacmackin** [comment]
Are we ready to merge now?

**Comment by will-saunders-ukaea** [comment]
> Are we ready to merge now?

It builds with gcc for me. I should do a intel build as well.

**Comment by oparry-ukaea** [comment]
Latest version builds with gcc and intel for me

**Comment by will-saunders-ukaea** [comment]
hipsycl cuda-nvcxx built and passed for me.

**Comment by will-saunders-ukaea** [comment]
Seeing this with oneapi@2022.1.0:
```
     166     [ 50%] Building CXX object test/CMakeFiles/unitTests.dir/unit/test_species.cpp.o
     167     cd /home/js0259/git-ukaea/NESO/spack-build-wdij2ez/test && /home/js0259/git/spack-v0.21/lib/spack/env/oneapi/icpx -DBOOST_HAS_NUMPY -DGTEST_LINKED_AS_SHARED_LIB
             RARY=1 -DH5_BUILT_AS_DYNAMIC_LIB -DNEKTAR_MEMORY_POOL_ENABLED -DNEKTAR_USE_MPI -pthread -DNEKTAR_USE_MPI -DNEKTAR_USE_SCOTCH -DNEKTAR_USE_THREAD_SAFETY -DNESO_I
             NTEL_MKL_FFT -DNESO_PARTICLES_DEVICE_TYPE_CPU -DNESO_PARTICLES_HDF5 -DRESTRICT=__restrict -DTIXML_USE_STL -DWITH_PYTHON -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_
             LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -D_POSIX_C_SOURCE=200809L -I/home/js0259/git-ukaea/NESO/include -isystem /opt/spack-v0.21/linux-ubuntu22.04-cascadelake/o
             neapi-2022.1.0/nektar-5.2.0-2022-09-03-fhauyh54iyvpvb4m7rsvxnx5gze3jqyg/include/nektar++ -isystem /opt/spack-v0.21/linux-ubuntu22.04-cascadelake/intel-2021.6.0/
             boost-1.83.0-s3rnueon7odwb3ltlvhokhrctnmzsnfa/include -isystem /opt/spack-v0.21/linux-ubuntu22.04-cascadelake/oneapi-2022.1.0/tinyxml-2.6.2-4mtob5q4ynkukm3wfzfy
             lo6262zdcpp3/include -isystem /opt/spack-v0.21/linux-ubuntu22.04-cascadelake/oneapi-2022.1.0/zlib-1.3-ydpab3sqqrc3drchckx3k2vtimkq325g/include -isystem /opt/spa
             ck-v0.21/linux-ubuntu22.04-cascadelake/oneapi-2022.1.0/intel-oneapi-mpi-2021.10.0-hf2xl2zozzfq77hj6gqgj6p2s5cvtamc/mpi/2021.10.0/include -isystem /opt/spack-v0.
             21/linux-ubuntu22.04-cascadelake/oneapi-2022.1.0/intel-oneapi-mpi-2021.10.0-hf2xl2zozzfq77hj6gqgj6p2s5cvtamc/include -isystem /opt/spack-v0.21/linux-ubuntu22.04
             -cascadelake/oneapi-2022.1.0/scotch-6.1.3-53r7376ur7shfppqtcmmtc4yp46cgskt/include -isystem /opt/spack-v0.21/linux-ubuntu22.04-cascadelake/oneapi-2022.1.0/fftw-
             3.3.10-xmexu2pv6ahkhub2yqlkpe2pe52fjjrn/include -isystem /opt/spack-v0.21/linux-ubuntu22.04-cascadelake/oneapi-2022.1.0/intel-oneapi-mkl-2023.2.0-7vwe4wkmwx5jvw
             yvdgxzplaj3aaa6rzg/mkl/2023.2.0/include -isystem /opt/spack-v0.21/linux-ubuntu22.04-cascadelake/oneapi-2022.1.0/neso-particles-working-5maus3isodrfkjqvqrxa3lsco
             vivu3vs/include -isystem /opt/spack-v0.21/linux-ubuntu22.04-cascadelake/oneapi-2022.1.0/hdf5-1.14.3-bslkajy4be5akrrscejlaapkj2kxhqjq/include -isystem /opt/spack
             -v0.21/linux-ubuntu22.04-cascadelake/oneapi-2022.1.0/googletest-1.12.1-plr6miqylajyh54v7uzi5aany34bdyfq/include --dpcpp -fsycl  -O3 -DNDEBUG -std=gnu++17 -DMKL_
             ILP64 -fsycl -MD -MT test/CMakeFiles/unitTests.dir/unit/test_species.cpp.o -MF CMakeFiles/unitTests.dir/unit/test_species.cpp.o.d -o CMakeFiles/unitTests.dir/un
             it/test_species.cpp.o -c /home/js0259/git-ukaea/NESO/test/unit/test_species.cpp
     168     /usr/bin/ld: libnesolib.a(mesh.cpp.o): in function `Mesh::sycl_solve_for_electric_field_fft(cl::sycl::queue&, FFT&)':
  >> 169     mesh-5db441.cpp:(.text+0x5219): undefined reference to `oneapi::mkl::dft::descriptor<(oneapi::mkl::dft::precision)36, (oneapi::mkl::dft::domain)32>::commit(cl::
             sycl::queue&)'
  >> 170     /usr/bin/ld: mesh-5db441.cpp:(.text+0x5273): undefined reference to `void oneapi::mkl::dft::compute_forward<oneapi::mkl::dft::descriptor<(oneapi::mkl::dft::prec
             ision)36, (oneapi::mkl::dft::domain)32>, std::complex<double>, std::complex<double> >(oneapi::mkl::dft::descriptor<(oneapi::mkl::dft::precision)36, (oneapi::mkl
             ::dft::domain)32>&, cl::sycl::buffer<std::complex<double>, 1, cl::sycl::detail::aligned_allocator<char>, void>&, cl::sycl::buffer<std::complex<double>, 1, cl::s
             ycl::detail::aligned_allocator<char>, void>&)'
  >> 171     /usr/bin/ld: mesh-5db441.cpp:(.text+0x55d9): undefined reference to `oneapi::mkl::dft::descriptor<(oneapi::mkl::dft::precision)36, (oneapi::mkl::dft::domain)32>
             ::commit(cl::sycl::queue&)'
  >> 172     /usr/bin/ld: mesh-5db441.cpp:(.text+0x5623): undefined reference to `void oneapi::mkl::dft::compute_backward<oneapi::mkl::dft::descriptor<(oneapi::mkl::dft::pre
             cision)36, (oneapi::mkl::dft::domain)32>, std::complex<double>, std::complex<double> >(oneapi::mkl::dft::descriptor<(oneapi::mkl::dft::precision)36, (oneapi::mk
             l::dft::domain)32>&, cl::sycl::buffer<std::complex<double>, 1, cl::sycl::detail::aligned_allocator<char>, void>&, cl::sycl::buffer<std::complex<double>, 1, cl::
             sycl::detail::aligned_allocator<char>, void>&)'
```

**Comment by will-saunders-ukaea** [comment]
Compiles and links outside the spack env with oneapi@2022.1.0 without issue.

**Comment by will-saunders-ukaea** [comment]
Update: main throws the same linking errors for me when using the spack env with oneapi@2022.1.0.

File: cmake/SYCL.cmake
Comment (JamesEdgeley) [inline cmake/SYCL.cmake:51]:
```suggestion
        function(add_sycl_to_target)
            cmake_parse_arguments(
                PARSE_ARGV 0
                ARGS ""
                "TARGET"
                "SOURCES"
            )
            set_property(SOURCE ${ARGS_SOURCES} APPEND PROPERTY COMPILE_OPTIONS "-fsycl")
            target_link_options(${ARGS_TARGET} PRIVATE "-fsycl")
        endfunction()
```

Might be missing something but shouldn't this function act like the hipsycl version? Possibly more flags necessary than suggested.

Comment (seimtpow) [reply cmake/SYCL.cmake:51]: IntelDPCPP doesn't need/define this function - having an empty version is just a hack to prevent some if/else clauses in the rest of the CMake files.

Comment (cmacmackin) [reply cmake/SYCL.cmake:51]: Most SYCL implementations define this function themselves. The only one that doesn't to my my knowledge is Intel and in that case the compiler doesn't need any additional flags to turn on SYCL.

**Comment by cmacmackin** [comment]
I've confirmed that the issue is with conflicting MKL versions. I'm going to see if I can modify the package file to prevent that from happening.

**Comment by cmacmackin** [comment]
The issue with linking for the OneAPI/MKL tool chain isn't really related to this PR. I've updated the NESO package to prevent it happening in future and will create separate PRs for that. In the meantime, I'm going to merge this.

**Pull request merged**
